pkg/controller/ecr/repository/controller.go

```diff
@@ -143,17 +153,37 @@ func (e *external) Observe(ctx context.Context, mgd resource.Managed) (managed.E
 		}
 	}
 
+	for _, awsClient := range e.subresourceClients {
```

Comment:
We used this pattern of subresources for S3 because it has a lot of them but I don't think we should do it for other resources unless we really have to. Do you think `LifecyclePolicy` can become a resource of its own?

And have you considered using [code generation](https://github.com/crossplane/provider-aws/blob/master/CODE_GENERATION.md) to generate it?

Reply:
I think that an ECR Repository has some similarities with S3 with that it has additional resources for each repository. SubResources like `RepositoryPolicy` and `LifecyclePolicy` only make sense in the context of the Repository, the way the api is structured. The intent of reusing the pattern from s3 were to reduce the complexity of the lifecycle functions, and handle some of the gocyclo warnings.

If the subresource pattern is not to be used here; do you have any input on how this should be implemented in a better way?

For Lifecycle policy at least, the API is a 1-1 relationship (https://docs.aws.amazon.com/sdk-for-go/api/service/ecr/#PutLifecyclePolicyInput). With a new resource specifically for LifecyclePolicy, it would always require a selector or ref to find another object. Information in the policy is also directly related to the repository (eg. the tag prefix).

TLDR; I dont think LifecyclePolicy should be a seperate resource. I started out with trying to use code gen for this, but ended up with this implementation.

Reply:
> For Lifecycle policy at least, the API is a 1-1 relationship (https://docs.aws.amazon.com/sdk-for-go/api/service/ecr/#PutLifecyclePolicyInput). With a new resource specifically for LifecyclePolicy, it would always require a selector or ref to find another object. Information in the policy is also directly related to the repository (eg. the tag prefix).

This is indeed not a rare pattern. For example, we have `IAMRolePolicyAttachment` where it always refers to a single `IAMRole` and a single `IAMPolicy` where it could've been that policy attachment could be a part of `IAMRole`. Another example could be [`VPCCIDRBlock`](https://doc.crds.dev/github.com/crossplane/provider-aws/ec2.aws.crossplane.io/VPCCIDRBlock/v1alpha1@v0.20.2) where it always targets a specific `VPC`. There are similar cases in other providers, too.

The way we usually think about such separations is not very absolute and you'll see exceptions here and there but having it as separate is preferred most of the time. One of the main reasons for this is the applicability of RBAC; if `LifecyclePolicy` is part of `Repository` then the restrictions you put in place cannot be as granular, everyone who can create `Repository` can create its `LifecyclePolicy` which might be fine but for example AWS lets you have that granularity and we're [trying to represent](https://github.com/crossplane/crossplane/blob/master/design/one-pager-managed-resource-api-design.md#high-fidelity) the API as close as possible.